### DISCUSSION TITLE
[levelup]: query string parser/stringifier

### DISCUSSION THREAD

**Discussion by 43081j** [e18e/e18e#1] [Ideas]
As part of cleanup (reducing footprint) and levelup (introducing modern, focused libraries), it makes sense to introduce a new high quality query string package.

That involves **parsing** and **stringifying**.

The recommended solution should really be to use `URLSearchParams` where possible. However, there are some reasons you may not be able to:

- Performance - some libraries can be faster than `URLSearchParams`
- Nested objects (e.g. `foo[bar]=123` should result in `{foo: {bar: 123}}`)
- Arrays (e.g. `foo[0]=1&foo[1]=2` should result in `{foo: [1, 2]}`)
- Custom delimiters (e.g. `foo=123;bar=456` instead of using `&`)
- Custom serializers - you may want to serialize a value a particular way when stringifying

## API

These are just proposed shapes of the API we provide to give an idea of what features we need.

Nothing is concrete, these could change when someone actually implements it.

### Shared types

```ts
type ArraySyntax =
  | 'index' // `foo[0]=a&foo[1]=b`
  | 'repeat'; // `foo[]=a&foo[]=b`

// This is a special return value for deserializers and serializers, to tell the library
// to fall back to using the default (de)serialize function.
// We can't just return `null` or `undefined` etc, because we may want to deserialize to that
declare const CONTINUE: Symbol;
```

### Parsing

```ts
declare function parse(
  input: string,
  options?: ParseOptions
): ParsedQuery;

type DeserializeValueFunction = (
  // This is the already deserialized key, which means if we're traversing `foo[bar]`,
  // the `key` will actually be `bar` with `nested` enabled
  key: string,
  value: string
) => unknown | (typeof CONTINUE);

type DeserializeKeyFunction = (
  // With nesting, this could be `foo[bar]`
  key: string
) => PropertyKey | (typeof CONTINUE); // We may want to return a non-string key, e.g. a number or symbol

interface ParseOptions {
  // Enable parsing nested objects and arrays
  // default: true
  nested: boolean;

  // Array syntax
  // default: "index"
  arraySyntax: ArraySyntax;

  // Delimiter to split kv pairs by
  // default: "&"
  delimiter?: string;

  // Custom deserializer
  deserializer?: DeserializeFunction;
}
```

### Stringifying

```ts
declare function stringify(
  input: unknown,
  options?: StringifyOptions
): string;

type SerializeValueFunction = (
  key: PropertyKey,
  value: unknown
) => string | (typeof CONTINUE);

type SerializeKeyFunction = (
  key: PropertyKey
) => string | (typeof CONTINUE); // We may want to return a non-string key, e.g. a number or symbol

interface StringifyOptions {
  // Enable parsing nested objects and arrays
  // default: true
  nested: boolean;

  // Array syntax
  // default: "index"
  arraySyntax: ArraySyntax;

  // Delimiter to split kv pairs by
  // default: "&"
  delimiter?: string;

 // Custom serializer
  serializer?: SerializeFunction;
}
```

### Parsing

```ts
parse('foo=1&bar=2');
// > `{foo: 1, bar: 2}`

parse('foo[0]=1&foo[1]=2');
// > `{foo: [1, 2]}`

parse('foo[]=1&foo[]=2', {arraySyntax: 'repeat'});
// > `{foo: [1, 2]}`

parse('foo[bar]=123');
// > `{foo: {bar: 123}}`

parse('foo[bar]=123', {nested: false});
// > `{'foo[bar]': 123}`

parse('foo=2010-01-01', {
  deserializer: (_key, val) => {
    // blindly try parse the val as a date
    const valAsDate = new Date(val);
    if (Number.isNaN(val.getDate())) {
      return CONTINUE; // Fall back to normal deserialization
    }
    return valAsDate; // Return the `Date`
  }
});
// > `{foo: Date}`
```

### Stringifying

```ts
stringify({foo: 123});
// > `"foo=123"`

stringify({foo: [1, 2]});
// > `"foo[0]=1&foo[1]=2"`

stringify({foo: {bar: 123}});
// > `"foo[bar]=123"`

stringify({foo: [1, 2]}, {arraySyntax: 'repeat'});
// > `"foo[]=1&foo[]=2"`

stringify({foo: {bar: 123}}, {nested: false});
// > `"foo=%5Bobject%20Object%5D"`
// In this one, we end up stringifying an object

stringify({foo: 123, bar: 456}, {delimiter: ';'});
// > `"foo=123;bar=456"`

stringify({foo: new Date('2000-01-01')}, {
  serializer: (_key, val) => {
    if (val instanceof Date) {
      return val.toISOString().split('T')[0];
    }
    return CONTINUE; // Fall back to default serialization
  }
});
// > `"foo=2000-01-01"`
```

## Implementation

This is a discussion in the umbrella org rather than an RFC in any particular project org because there's no preference here of who owns or creates this.

The ecosystem seems to be lacking such a package right now, which means it is difficult to improve many projects since there are no lighter alternatives to what they already use.

If there _is_ such a package, it may turn out we're just not aware and we should be pushing it forwards as a community.

If there is _not_ such a package, we need an **owner** to run with this (wherever) and i, amongst others, would likely be open to helping out with contributions.

### Re-use where possible

`fast-querystring` and `URLSearchParams` in particular are already very well tested, fast methods of parsing and stringifying basic query strings (no nesting, no arrays, etc).

If we implement this package, it may make sense to use one of those under the hood rather than having to implement the same algorithms ourselves.

That may not be possible if we allow customisation of delimiters, but maybe we can contribute upstream to `fast-querystring` to make that an option (or we just drop the delimiter option).

## Prior art

| name | nesting | arrays | package size | notes |
| - | - | - | - | - |
| `node:querystring` | ❌ | ❌ | 0KB / 0 packages | built in to node |
| `fast-querystring` | ❌ | ❌ | 28KB / 2 packages | possibly fastest library right now |
| `URLSearchParams` | ❌ | ❌ | 0KB / 0 packages | built-in to all platforms |
| `qs` | ✅ | ✅ | 592KB / 15 packages | large footprint, much slower |
| `nested-querystring` | ✅ | ✅ | 7.2KB / 1 package | small, not configurable |
| `querystringparser` | ✅ | ✅  | 26KB / 1 package | old, not maintained |

### Existing packages which satisfy some of the requirements already

These are packages which almost satisfy the requirements above. So it may make sense to contribute to them rather than building a new one.

- [faster-qs](https://github.com/hans00/faster-qs) - uses fast-querystring under the hood

**Comment by dominikg** [comment]
re array syntax, what about foo=1&foo=2 (like searchparams `new URLSearchParams('foo=1&foo=2').getAll('foo')`) or split at delimiter foo=1~2 (delimiter char needs to be rare and url safe, tilde is one of the best i know) ? not repeating the key is important for arrays with many values to save chars. even more can be saved when array values are a known finite set by turning it into a bitmask.


With `getAll` i'd say URLSearchParams does support arrays.
 
A way to define per key parse/serialize functions would go a long way to allow users to keep urls short.

Another thing to have is stable sorting. `?foo=1&bar=2` is not the same as `?bar=2&foo=1` for caching.
